### PR TITLE
CLD-5704 Migrate daily master/cloud tests

### PR DESCRIPTION
**Pull request by mvitale1989**
#### Summary

This PR adds support for daily master/cloud E2E tests in GHA.
- The `REPORT_TYPE` input will decide how the test will behave.
- In case of daily tests (`MASTER`, `MASTER_UNSTABLE`, `CLOUD` and `CLOUD_UNSTABLE`), the `mattermostdevelopment/mattermost-enterprise-edition:master` image is automatically picked, no matter the given `commit_sha` or `PR_NUMBER` you specified at triggering time
- Zephyr reporting is implemented

The triggering of this job will be done by argo-events.

#### Ticket Link

https://mattermost.atlassian.net/browse/CLD-5704

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by mvitale1989** [comment]
:point_up: the run above was triggered for the `CLOUD_UNSTABLE` test, and produced the following message in the channel: https://community.mattermost.com/core/pl/7pn9rcgk1iygirz5zqx8ntnwmw

**Comment by mvitale1989** [comment]
:point_up: the run above was triggered for the `CLOUD` test, and it uses the latest `master` image available.

**Review by saturninoabril — APPROVED** [review]
Love it, great work @mvitale1989!

File: .github/workflows/e2e-fulltests-ci.yml [hunk 1]
```diff
@@ -99,19 +114,36 @@ jobs:
...
+          SERVER=onprem
+          case "$REPORT_TYPE" in
             NONE | PR)
               echo "status_check_context=E2E Tests/test${MM_ENV:+/$MM_ENV_HASH}" >> $GITHUB_OUTPUT
               echo "workers_number=20" >> $GITHUB_OUTPUT
               echo "ENABLED_DOCKER_SERVICES=postgres inbucket minio openldap elasticsearch keycloak" >> $GITHUB_OUTPUT
               echo "TEST_FILTER=$TEST_FILTER_PR" >> $GITHUB_OUTPUT
               echo "BUILD_ID=${BUILD_ID_PREFIX}-pr-onprem-ent" >> $GITHUB_OUTPUT
               ;;
+            MASTER | MASTER_UNSTABLE | CLOUD | CLOUD_UNSTABLE)
```
[the remark this answers is not in the document]
Comment (mvitale1989) [reply .github/workflows/e2e-fulltests-ci.yml:126]: I agree that the logic we're migrating is more complex than it should. I have already collected some points to rework the abstraction offered by the workflow and simplify the scripts, I'll create a ticket for that :+1: 

In the meantime i'd propose simply mirroring what we have to ensure it's working fine on GHA, before going ahead with the improvements. Wdyt?

[the remark this answers is not in the document]
Comment (saturninoabril) [reply .github/workflows/e2e-fulltests-ci.yml:126]: Thanks and yes, totally agree.

File: e2e-tests/.ci/report.publish.sh [hunk 2]
```diff
@@ -10,40 +10,65 @@ cd "$(dirname "$0")"
...
+  if ! grep -qE "${BRANCH_REGEX}"<<<"$BRANCH"; then
+    mme2e_log "Error: when using TYPE=PR, the BRANCH variable should respect regex '$BRANCH_REGEX'. Aborting." >&2
+    exit 1
+  fi
+  export PULL_REQUEST="https://github.com/mattermost/mattermost/pull/${BRANCH##*-}"
+fi
 
 # Env vars used during the test. Their values will be included in the report
-: ${BRANCH:?}
+: ${BRANCH:?} # May be either a ref, a commit hash, or 'server-pr-PRNUMBER' (if TYPE=PR)
```
[the remark this answers is not in the document]
Comment (mvitale1989) [reply e2e-tests/.ci/report.publish.sh]: Fixed in this commit: https://github.com/mattermost/mattermost/pull/27393/commits/e8e74d2c11c0ba2d06664a9781cb0dae330bb766
Along with other fixes and improvements (e.g. zephyr's folderIds were wrong for unstable tests)

File: .github/workflows/e2e-fulltests-ci.yml [hunk 3]
```diff
@@ -99,19 +114,35 @@ jobs:
...
+              ;;
+            MASTER | MASTER_UNSTABLE | CLOUD | CLOUD_UNSTABLE)
+              IS_TEST_UNSTABLE=$(sed -n -E 's/^.*(UNSTABLE).*$/\1/p' <<< "$REPORT_TYPE") # Variable is UNSTABLE if report type is for unstable tests
+              if grep -q CLOUD <<<$REPORT_TYPE; then
+                SERVER=cloud
+              fi
+              TEST_FILTER_VARIABLE="TEST_FILTER_DAILY_${SERVER@U}"
+              BUILD_ID_SUFFIX="${IS_TEST_UNSTABLE:+unstable-}daily-${SERVER}-ent"
+              echo "status_check_context=E2E Tests/test-${BUILD_ID_SUFFIX}${MM_ENV:+/$MM_ENV_HASH}" >> $GITHUB_OUTPUT
+              echo "workers_number=10" >> $GITHUB_OUTPUT # Daily tests are not time critical, and it's more efficient to run on fewer workers
```
Comment (saturninoabril) [inline .github/workflows/e2e-fulltests-ci.yml:133]: 👍
(Originally, all 20 workers except for unstable test but sounds good for daily tests.)

File: e2e-tests/cypress/save_report.js [hunk 4]
```diff
@@ -24,7 +24,9 @@
  *   For sending hooks to Mattermost channels
  *      - FULL_REPORT, WEBHOOK_URL and DIAGNOSTIC_WEBHOOK_URL
  *   Test type
- *      - TYPE=[type], e.g. "MASTER", "PR", "RELEASE", "CLOUD"
+ *      - TYPE=[type]; valid values: "PR", "RELEASE", "MASTER", "MASTER_UNSTABLE", "CLOUD", "CLOUD_UNSTABLE", "NONE"
```
Comment (toninis) [inline e2e-tests/cypress/save_report.js:27]: We do not need cloud anymore correct ?

**Review by toninis — APPROVED** [review]
I would love to see this integrated with spinwicks on demand 🚀

**Comment by mvitale1989** [comment]
> I would love to see this integrated with spinwicks on demand 🚀

Thanks for reviewing :heart: as of now the tests depend on accessing the dependencies directly (e.g. we need to check if some migrations are complete), but it's worth checking if we can adapt them to run against spinwicks. I'll create a ticket to explore this

**Pull request merged**